lib/lltz_michelson/instruction.ml

```diff
@@ -0,0 +1,562 @@
+              "Instruction.Slot.def_all: invalid slots"
+                (stack : Stack.t)
+                (slots' : [< Stack.Slot.definable ] list)]
+      in
+      match List.zip_with_remainder stack slots' with
+      | slots, None ->
+        check_slots slots;
+        let slots = (slots :> (Stack.Slot.t * Stack.Slot.t) list) in
+        List.map ~f:snd slots
+      | slots, Some (First right) ->
+        check_slots slots;
+        let slots = (slots :> (Stack.Slot.t * Stack.Slot.t) list) in
+        List.map ~f:snd slots @ right
+      | _ ->
+        raise_s [%message "Instruction.Slot.def_all: invalid stack" (stack : Stack.t)]
+    in
+    in_ stack
+  ;;
+
+  let collect_all slots = seq (List.map slots ~f:collect)
+  let let_all slots ~in_ = seq [ def_all slots ~in_; collect_all slots ]
+
+  let lookup slot ~in_ stack =
+    let idx = Stack.find_exn stack slot in
+    in_ idx stack
+  ;;
+
+  let dup slot = lookup slot ~in_:(fun idx -> dup idx)
+
+  let set slot stack =
+    lookup slot stack ~in_:(fun idx stack ->
+        let instructions =
+          match idx with
+          | 0 -> assert false
+          | n -> I.[ dug_n n; dig_n (n - 1); drop ]
+        in
+        match List.split_n stack idx with
+        | _val_ :: left, nth :: right -> Config.ok (left @ (nth :: right)) instructions
+        | _ ->
+          raise_s
+            [%message
+              "Instruction.Slot.set: invalid stack"
+                (stack : Stack.t)
+                (slot : [< Stack.Slot.definable ])])
+  ;;
+end
+
+let dip n (t : t) : t =
+ fun stack ->
+  let left_stack, right_stack = List.split_n stack n in
+  let { Config.stack = right_stack; instructions } = t right_stack in
+  let stack =
+    Config.Stack.apply right_stack ~f:(fun right_stack -> left_stack @ right_stack)
+  in
+  { Config.stack; instructions = [ I.dip_n n instructions ] }
+;;
+
+let unpair_n n =
+  (* invariant: pop 1 value, push n *)
+  match n with
+  | 0 -> drop 1
+  | 1 -> noop
+  | n -> prim 1 n (I.unpair_n n)
+;;
+
+let pair_n n =
+  (* invariant: pop n values, push 1 *)
+  match n with
+  | 0 -> prim 0 1 I.unit
+  | 1 -> noop
+  | n -> prim n 1 (I.pair_n n)
+;;
+
+let get_n idx ~length:n =
+  match n with
+  | 0 ->
+    raise_s [%message "Comp.Instruction.tuple_project: cannot project from empty tuple"]
+  | 1 ->
+    assert (idx = 0);
+    noop
+  | n ->
+    (* n >= 2 *)
+    let k = if idx = n - 1 then 2 * idx else (2 * idx) + 1 in
+    prim 1 1 (I.get_n k)
+;;
+
+let update_n idx ~length:n =
+  (* expected stack: value, tuple, ... *)
+  (* result stack: tuple' *)
+  match n with
+  | 0 -> raise_s [%message "Com.Instruction.tuple_update: cannot update the empty tuple"]
+  | 1 ->
+    assert (idx = 0);
+    remove 1
+  | n ->
+    let k = if idx = n - 1 then 2 * idx else (2 * idx) + 1 in
+    prim 2 1 (I.update_n k)
+;;
+
+(*  Lambdas used in LLTZ-IR do not use heaps. *)
+let lambda  ~lam_var ~return_type return stack =
+  let parameter_slot = (match lam_var with
```

Comment:
437: let parameter_slot = `Ident (fst lam_var) in
439: let parmeter_type = snd lam_var in